server/src/test/java/org/elasticsearch/transport/TransportServiceLifecycleTests.java

```diff
@@ -165,6 +170,31 @@ public void testInternalSendExceptionForksToGenericIfHandlerDoesNotFork() {
 
             assertEquals("simulated exception in sendRequest", getSendRequestException(future, IOException.class).getMessage());
         }
+        assertWarnings(
+            "[transport.enable_stack_protection] setting was deprecated in Elasticsearch and will be removed in a future release."
+        );
+    }
+
+    public void testInternalSendExceptionWithNonForkingResponseHandlerCanCompleteFutureWaiterFromGenericThread() throws Exception {
+        try (var nodeA = new TestNode("node-A")) {
+            final var future = new PlainActionFuture<TransportResponse.Empty>();
+            final var latch = new CountDownLatch(1);
+            nodeA.transportService.getThreadPool().generic().execute(() -> {
+                assertEquals("simulated exception in sendRequest", getSendRequestException(future, IOException.class).getMessage());
+                latch.countDown();
+            });
+            nodeA.transportService.sendRequest(
+                nodeA.getThrowingConnection(),
+                TestNode.randomActionName(random()),
+                new EmptyRequest(),
+                TransportRequestOptions.EMPTY,
+                new ActionListenerResponseHandler<>(future.delegateResponse((l, e) -> {
+                    assertThat(Thread.currentThread().getName(), startsWith("TEST-"));
+                    l.onFailure(e);
+                }), unusedReader(), EsExecutors.DIRECT_EXECUTOR_SERVICE)
+            );
+            assertBusy(() -> assertTrue(future.isDone()));
```

Comment:
Maybe just `safeGet(future)`? No need to busy-wait on the completion of a future, we can just block the thread directly.

Reply:
The future throws so that `safeGet` does not work here. Maybe we could add a `safeGetFailure` utility method?

Reply:
Ah yes I see. Still it'd be better to use `future.get()` to avoid the busy-wait and then catch whatever we expect it to throw.